File: flutter/packages/workout_log_core/lib/src/io/session_storage.dart
/// The byte-level backend a [SessionStore] sits on.
///
/// Declared here, implemented per platform: this is the seam that keeps
/// `dart:io` (and the browser, which has no filesystem at all) out of the core.
/// An id is a bare filename such as `2026-08-06_D2.json`.
abstract interface class SessionStorage {
  Future<List<String>> listIds();

  Future<String> read(String id);

  Future<void> write(String id, String contents);

  Future<void> delete(String id);

  /// A human-readable name for the backing location, for the status bar.
  String get label;
}

/// In-memory backend: the whole store for web, and the test double everywhere else.
class MemoryStorage implements SessionStorage {
  MemoryStorage({Map<String, String>? seed, this.label = 'in-memory'})
      : _files = {...?seed};

  final Map<String, String> _files;

  @override
  final String label;

  Map<String, String> get snapshot => Map.unmodifiable(_files);

  @override
  Future<List<String>> listIds() async => _files.keys.toList();

  @override
  Future<String> read(String id) async {
    final contents = _files[id];
    if (contents == null) throw StateError('no such session: $id');
    return contents;
  }

  @override
  Future<void> write(String id, String contents) async => _files[id] = contents;

  @override
  Future<void> delete(String id) async => _files.remove(id);
}

File: crates/workout-log-core/src/io/session_storage.rs
//! The byte-level backend a [`super::session_store::SessionStore`] sits on.
//!
//! Declared here, implemented per platform: this is the seam that keeps the
//! filesystem out of the core. An id is a bare filename, `2026-08-06_D2.json`.

use std::collections::BTreeMap;

#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
#[error("{0}")]
pub struct StorageError(pub String);

pub type StorageResult<T> = Result<T, StorageError>;

pub trait SessionStorage: Send + Sync {
    fn list_ids(&self) -> StorageResult<Vec<String>>;
    fn read(&self, id: &str) -> StorageResult<String>;
    fn write(&mut self, id: &str, contents: &str) -> StorageResult<()>;
    fn delete(&mut self, id: &str) -> StorageResult<()>;
    /// A human-readable name for the backing location, for the status bar.
    fn label(&self) -> String;
}

/// In-memory backend: the test double, and the whole store wherever there is no
/// folder to point at.
#[derive(Debug, Clone)]
pub struct MemoryStorage {
    files: BTreeMap<String, String>,
    label: String,
}

impl Default for MemoryStorage {
    fn default() -> Self {
        Self {
            files: BTreeMap::new(),
            label: "in-memory".into(),
        }
    }
}

impl MemoryStorage {
    pub fn new(label: impl Into<String>) -> Self {
        Self {
            files: BTreeMap::new(),
            label: label.into(),
        }
    }

    pub fn seeded(seed: BTreeMap<String, String>, label: impl Into<String>) -> Self {
        Self {
            files: seed,
            label: label.into(),
        }
    }

    pub fn snapshot(&self) -> &BTreeMap<String, String> {
        &self.files
    }
}

impl SessionStorage for MemoryStorage {
    fn list_ids(&self) -> StorageResult<Vec<String>> {
        Ok(self.files.keys().cloned().collect())
    }

    fn read(&self, id: &str) -> StorageResult<String> {
        self.files
            .get(id)
            .cloned()
            .ok_or_else(|| StorageError(format!("no such session: {id}")))
    }

    fn write(&mut self, id: &str, contents: &str) -> StorageResult<()> {
        self.files.insert(id.to_string(), contents.to_string());
        Ok(())
    }

    fn delete(&mut self, id: &str) -> StorageResult<()> {
        self.files.remove(id);
        Ok(())
    }

    fn label(&self) -> String {
        self.label.clone()
    }
}
